docs/ql-libraries/dataflow/dataflow.md

```diff
@@ -0,0 +1,472 @@
+# Using the shared data-flow library
+
+This document is aimed towards language maintainers and contain implementation
+details that should be mostly irrelevant to query writers.
+
+## Overview
+
+The shared data-flow library implements sophisticated global data flow on top
+of a language-specific data-flow graph. The language-specific bits supply the
+graph through a number of predicates and classes, and the shared implementation
+takes care of matching call-sites with returns and field writes with reads to
+ensure that the generated paths are well-formed. The library also supports a
+number of additional features for improving precision, for example pruning
+infeasible paths based on type information.
+
+## File organisation
+
+The data-flow library consists of a number of files typically located in
+`<lang>/dataflow` and `<lang>/dataflow/internal`:
+
+```
+dataflow/DataFlow.qll
+dataflow/internal/DataFlowImpl.qll
+dataflow/internal/DataFlowCommon.qll
+dataflow/internal/DataFlowImplSpecific.qll
+```
+
+`DataFlow.qll` provides the user interface for the library and consists of just
+a few lines of code importing the implementation:
+
+#### `DataFlow.qll`
+```ql
+import <lang>
+
+module DataFlow {
+  import semmle.code.java.dataflow.internal.DataFlowImpl
+}
+```
+
+The `DataFlowImpl.qll` and `DataFlowCommon.qll` files contain the library code
+that is shared across languages. These contain `Configuration`-specific and
+`Configuration`-independent code, respectively. This organization allows
```

Comment:
allows for?

Reply:
Actually the sentence was just missing something at the end. Fixed.